[main] mediatek: filogic: add Unielec U7986-06 (EMMC/NAND)(AX6000)

ZBXallen opened the pull request:
Hardware specification
----------------------
SoC: MediaTek MT7986A 4x A53
Flash: 128MB SPI-NAND and 8GB eMMC
RAM: 1GB DDR4
Ethernet:
  * WAN:1x 2.5GbE (Airoha EN8811H)
  * LAN:4x 10/100/1000 Mbps (Switch: MediaTek MT7531AE) WIFI: Mediatek MT7975PN+MT7975N 802.11ax 2.4/5 GHz 4T4R(AX6000) Interfaces:
  * M.2 Key-B: USB 3.0 with SIM slot
  * front USB 2.0 port LED: Power, Status, WLAN2G, WLAN5G, LTE, USER1, USER2 Button: Reset
Power: 12V DC-005-T20
UART: 3.3v, 115200n8
VCC RX TX GND

The U-boot menu will automatically appear at startup, and then select the required options through UP/DOWN Key. NAND Flash and eMMC Flash instructions:
1. Set your computers IP adress to 192.168.1.2.
2. Run a TFTP server providing the sysupgrade.bin image.
3. Power on the router, into the U-Boot menu.
4. Select "2. Upgrade firmware"
5. Update sysupgrade.bin file name, input server IP and input device IP (if they deviate from the defaults)
6. Wait for automatic startup after burning

Signed-off-by: Allen Zhao <[allenzhao@unielecinc.com](mailto:allenzhao@unielecinc.com)>

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+		};
+	};
+
+	wf_dbdc_pins: wf_dbdc-pins {
```
I thought the wf_dbdc_pins were not needed for ax6000 devices?

ZBXallen replied:
I think it may need to be added, according to the default configuration.

aiamadeus replied:
If you remove this and wifi still works, it's not needed.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
```
It's better to add a blank line here.

ZBXallen replied:
OK,it has been fixed.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
```
It's better to add a blank line here.

ZBXallen replied:
OK,it has been fixed.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+		led_status: led@0 {
+			label = "sys:green";
+			gpios = <&pio 21 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
`default-state = "off";` is not needed.

ZBXallen replied:
OK,it has been fixed.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06-emmc.dts:
```diff
@@ -0,0 +1,57 @@
+ */
+
+/dts-v1/;
+#include "mt7986a-unielec-u7986-06.dtsi"
```
It's better to add a blank line here.

ZBXallen replied:
OK

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06-nand.dts:
```diff
@@ -0,0 +1,71 @@
+ */
+
+/dts-v1/;
+#include "mt7986a-unielec-u7986-06.dtsi"
```
It's better to add a blank line here.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+		regulator-always-on;
+	};
+
+	usb_vbus: regulator-usb-vbus {
```
Not used for the ssusb node?

ZBXallen replied:
Yes, I have deleted it.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+		switch@0 {
+			compatible = "mediatek,mt7531";
+			reg = <31>;
+			reset-gpios = <&pio 5 0>;
```
`reset-gpios = <&pio 5 GPIO_ACTIVE_HIGH>;`
Missing interrupt properties?

ZBXallen replied:
OK, I have added it now.
Please check, thanks.

hauke replied:
You did not add it.

ZBXallen replied:
Minor mistake, I fixed it, please check it, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+	model = "Unielec U7986-06";
+	compatible = "unielec,u7986-06", "mediatek,mt7986";
+
+
```
Extra blank lines.

ZBXallen replied:
OK,it has been fixed.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+	status = "okay";
+};
+
+
```
Extra blank lines.

ZBXallen replied:
OK,it has been fixed.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+	};
+};
+
+
```
Extra blank lines.

aiamadeus commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -118,6 +118,9 @@ mediatek_setup_interfaces()
 	xiaomi,redmi-router-ax6000-ubootmod)
 		ucidef_set_interfaces_lan_wan "lan2 lan3 lan4" wan
 		;;
+	unielec,u7986-06*)
```
Please keep alphabetical order.
BTW, does your device shell mark start from lan0?

ZBXallen replied:
Yes, I have sorted them in alphabetical order now. Additionally, the device shell starts from lan1.
Please check, thanks.

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+			reset-assert-us = <10000>;
+			reset-deassert-us = <20000>;
+
+			phy-mode = "2500base-x";
```
OK, it has been fixed.

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+
+	gmac0: mac@1 {
+		compatible = "mediatek,eth-mac";
+		reg = <1>;
```
OK, it has been fixed.
Please check, thanks.

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+		};
+/*
+		phy6: phy@6 {
+			compatible = "ethernet-phy-id67c9.de0a";
```
OK, it has been fixed.

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+
+&gmac1 {
+        phy-mode = "2500base-x";
+        phy-connection-type = "2500base-x";
```
OK, it has been fixed.
Please check, thanks.

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+
+	i2c_pins: i2c-pins-3-4 {
+			mux {
+					function = "i2c";
```
OK, it has been fixed.
Please check, thanks.

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+	status = "okay";
+
+	proslic_spi: proslic_spi@0 {
+		compatible = "silabs,proslic_spi";
```
Yes, this is an extra thing.

aiamadeus replied:
This looks to be copied directly from the sdk.

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,522 @@
+	status = "okay";
+
+	wm8960: wm8960@1a {
+		compatible = "wlf,wm8960";
```
Yes, this is an extra thing.

ZBXallen commented:
@aiamadeus Hello Aiamadeus, sorry for the late reply.
Is it OK？Please check. Thanks

hauke commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,506 @@
+	aliases {
+		ethernet0 = &gmac0;
+		ethernet1 = &gmac1;
+		serial0 = &uart0;
```
You can add this to define the label_mac as `gmac1` here:
```
label-mac-device = &gmac1;
```

ZBXallen replied:
OK, I have added it now.
Please check, thanks.

hauke commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,506 @@
+		compatible = "gpio-leds";
+
+		led_status: led@0 {
+			label = "sys:green";
```
Please use the color and function attribute:
```suggestion
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_STATUS;
```
Same for the other LEDs.

Full list here: https://github.com/torvalds/linux/blob/master/include/dt-bindings/leds/common.h

ZBXallen replied:
OK, I have added it now.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,506 @@
+
+				port@0 {
+					reg = <0>;
+					label = "lan0";
```
No longer needed?

ZBXallen replied:
Yes,it`s only lan1-lan4, port 0,1,3,4. I have added it now.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06-emmc.dts:
```diff
@@ -0,0 +1,58 @@
+	compatible = "unielec,u7986-06-emmc", "mediatek,mt7986";
+};
+
+&emmc_partitions {
```
I think this would be better?
emmc partitions are not used elsewhere
```
&mmc0 {
	card@0 {
		compatible = "mmc-card";
		reg = <0>;

		block {
			compatible = "block-device";

			partitions {
			......
			};
		};
	};
};
```

ZBXallen replied:
OK, I have added it now.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06-nand.dts:
```diff
@@ -0,0 +1,72 @@
+	compatible = "unielec,u7986-06-nand", "mediatek,mt7986";
+};
+
+&nand_partitions {
```
I think this would be better?
nand partitions are not used elsewhere
```
&spi0 {
	spi_nand@0 {
		mediatek,nmbm;
		mediatek,bmt-max-ratio = <1>;
		mediatek,bmt-max-reserved-blocks = <64>;

		partitions {
			.......
		};
	};
};
```

ZBXallen replied:
OK, I have added it now.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,506 @@
+
+		block {
+			compatible = "block-device";
+			emmc_partitions: partitions {
```
This should be placed in emmc.dts

ZBXallen replied:
OK, I have added it now.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,506 @@
+		mediatek,bmt-max-ratio = <1>;
+		mediatek,bmt-max-reserved-blocks = <64>;
+
+		nand_partitions: partitions {
```
This should be placed in nand.dts

ZBXallen replied:
OK, I have added it now.
Please check, thanks.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,451 @@
+		};
+	};
+
+	wf_dbdc_pins: wf_dbdc-pins {
```
Can you try removing wf_dbdc_pins and check if wifi works?

ZBXallen replied:
OK, it has been fixed.
Please check, thanks.

aiamadeus replied:
The pinctrl-names of wifi also need to remove dbdc.

ZBXallen replied:
OK, it has been fixed.

aiamadeus replied:
The wifi node has not changed?

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06-nand.dts:
```diff
@@ -0,0 +1,95 @@
+				reg = <0x0100000 0x0080000>;
+			};
+
+			factory: partition@180000 {
```
OK, I have added it now.
Please check, thanks.

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,451 @@
+	pinctrl-0 = <&i2c_pins>;
+	status = "okay";
+
+	wm8960: wm8960@1a {
```
There is no chip, but the interface exists, it has been fixed.
Please check, thanks.

ZBXallen commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,451 @@
+	pinctrl-0 = <&spic_pins_g2>;
+	status = "okay";
+
+	proslic_spi: proslic_spi@0 {
```
There is no chip, but the interface exists, it has been fixed.
Please check, thanks.

ZBXallen commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -203,6 +204,10 @@ mediatek_setup_macs()
 		lan_mac=$(macaddr_add "$label_mac" 1)
 		;;
+	unielec,u7986-06*)
+		wan_mac=$(macaddr_add $(cat /sys/class/net/eth0/address) 1)
```
OK, it has been fixed.
Please check, thanks.

aiamadeus replied:
From `...lan_wan "lan1 lan2 lan3 lan4" eth1`
Update dts so there is no need to set mac here:
```
wan_mac -> eth1 mac
lan_mac -> eth0 mac
```

ZBXallen replied:
So, I don't need to add code in “mediatek_setup_macs()”?

ZBXallen replied:
OK,Fixing now, thanks.

ZBXallen committed:
mediatek: filogic: add Unielec U7986-06 (EMMC/NAND)(AX6000)
Hardware specification
----------------------
SoC: MediaTek MT7986A 4x A53
Flash: 128MB SPI-NAND and 8GB eMMC
RAM: 1GB DDR4
Ethernet:
  * WAN:1x 2.5GbE (Airoha EN8811H)
  * LAN:4x 10/100/1000 Mbps (Switch: MediaTek MT7531AE)
WIFI: Mediatek MT7975PN+MT7975N 802.11ax 2.4/5 GHz 4T4R(AX6000)
Interfaces:
  * M.2 Key-B: USB 3.0 with SIM slot
  * front USB 2.0 port
LED: Power, Status, WLAN2G, WLAN5G, LTE, USER1, USER2
Button: Reset
Power: 12V DC-005-T20
UART: 3.3v, 115200n8
VCC RX TX GND

The U-boot menu will automatically appear at startup, and then select the required options through UP/DOWN Key.
NAND Flash and eMMC Flash instructions:
1. Set your computers IP adress to 192.168.1.2.
2. Run a TFTP server providing the sysupgrade.bin image.
3. Power on the router, into the U-Boot menu.
4. Select "2. Upgrade firmware"
5. Update sysupgrade.bin file name, input server IP and input device IP (if they deviate from the defaults)
6. Wait for automatic startup after burning

Signed-off-by: Allen Zhao <allenzhao@unielecinc.com>

hauke commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,416 @@
+					label = "lan4";
+				};
+
+				port@5 {
+					reg = <5>;
+					label = "lan5";
+					phy-mode = "2500base-x";
+
+					fixed-link {
+						speed = <2500>;
+						full-duplex;
+						pause;
+					};
+				};
```
Why do you configure this to fixed link? 
If this port is not used just do not configure if here.

hauke commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,416 @@
+		};
+		conf-pd {
+			pins = "SPI2_CLK", "SPI2_MOSI", "SPI2_MISO";
+			drive-strength = <8>;
```
There is a discussion to lower SPI drive strength from 8mA to 4mA when using 52MHz SPI clock, see https://github.com/openwrt/openwrt/pull/18752
MTK did this in a recent SDK change, see https://git01.mediatek.com/plugins/gitiles/openwrt/feeds/mtk-openwrt-feeds/+/003744197aa3a587828b4330ab1112ebdb9e840a%5E%21/

Maybe it is useful for you too.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,416 @@
+		stdout-path = "serial0:115200n8";
+	};
+
+	memory {
```
```suggestion
	memory@40000000 {
		reg = <0 0x40000000 0 0x40000000>;
		device_type = "memory";
	};
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,416 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led_status: led@0 {
```
Please use `led-[0-9]*`.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,416 @@
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_INDICATOR;
+			gpios = <&pio 17 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
Please remove all `default-state = "off"`. This is the default status on Linux.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,416 @@
+			};
+		};
+
+		switch@0 {
```
```suggestion
		switch@1f {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,416 @@
+			       "EMMC_DATA_3", "EMMC_DATA_4", "EMMC_DATA_5",
+			       "EMMC_DATA_6", "EMMC_DATA_7", "EMMC_CMD";
+			input-enable;
+			drive-strength = <4>;
```
```suggestion
			drive-strength = <MTK_DRIVE_4mA>;
```
ref: f3ce0c29ea0ec56b702a38f3aca84fb6648008ba

DragonBluep commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,416 @@
+			       "EMMC_DATA_6", "EMMC_DATA_7", "EMMC_CMD";
+			input-enable;
+			drive-strength = <4>;
+			mediatek,pull-up-adv = <1>;	/* pull-up 10K */
```
Please use `bias-pull-up`. ref: c089ad8f68ffdcc45ee6fd856bb92266d0a2f2c4

DragonBluep commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06.dtsi:
```diff
@@ -0,0 +1,416 @@
+		};
+	};
+
+	mmc0_pins_uhs: mmc0-uhs-pins {
```
This pinctrl group seems useless. It's same as `mmc0_pins_default`.

DragonBluep commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1697,6 +1697,29 @@ define Device/unielec_u7981-01-nand
+  DEVICE_MODEL := U7986-06
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-hwmon-pwmfan kmod-mt7915e kmod-mt7986-firmware kmod-phy-airoha-en8811h \
+		     kmod-usb3 e2fsprogs f2fsck mkf2fs mt7986-wo-firmware
```
pwm and filesystem packages are unnecessary. It seems that sysupgrade config is missing?

DragonBluep commented on target/linux/mediatek/dts/mt7986a-unielec-u7986-06-emmc.dts:
```diff
@@ -0,0 +1,84 @@
+ * Copyright (C) 2023 Allen Zhao <allenzhao@unielecinc.com>
+ */
+
+/dts-v1/;
```
The duplicated `/dts-v1/;` can be removed from device .dts files.
